src/io/metadata.cpp

```diff
@@ -58,6 +58,33 @@ void Metadata::Init(data_size_t num_data, int weight_idx, int query_idx) {
   }
 }
 
+void Metadata::InitByReference(data_size_t num_data, const Metadata* reference) {
+  int has_weights = reference->num_weights_ > 0;
+  int has_init_scores = reference->num_init_score_ > 0;
+  int has_queries = reference->num_queries_ > 0;
+  int nclasses = reference->num_classes();
+  Init(num_data, has_weights, has_init_scores, has_queries, nclasses);
+}
+
+void Metadata::Init(data_size_t num_data, int32_t has_weights, int32_t has_init_scores, int32_t has_queries, int32_t nclasses) {
+  num_data_ = num_data;
+  label_ = std::vector<label_t>(num_data_);
+  if (has_weights) {
+    weights_ = std::vector<label_t>(num_data_, 0.0f);
+    num_weights_ = num_data_;
+    weight_load_from_file_ = false;
+  }
+  if (has_init_scores) {
+    num_init_score_ = static_cast<int64_t>(num_data) * nclasses;
+    init_score_ = std::vector<double>(num_init_score_, 0);
+  }
+  if (has_queries) {
+    if (!query_weights_.empty()) { query_weights_.clear(); }
+    queries_ = std::vector<data_size_t>(num_data_, 0);
```

Comment:
A better way to do this should be clear and resize `queries_` directly, to avoid unnecessary memory allocation and reallocation?

Reply:
I just copied this from the existing Init() method, but happy to change it.  What about just throwing an error if there is data already?  This is really an unexpected scenario... calling Init() after having already called it once or pushed some data.

Reply:
Sure. I think that works.